File: src/com/boardgamegeek/provider/IndirectFileProvider.java
package com.boardgamegeek.provider;

import android.content.Context;
import android.net.Uri;
import android.text.TextUtils;

import com.boardgamegeek.database.ResolverUtils;

public abstract class IndirectFileProvider extends BaseFileProvider {
	protected abstract Uri getFileUri(Uri uri);

	protected abstract String getColumnName();

	@Override
	protected String generateFileName(Context context, Uri uri) {
		String url = ResolverUtils.queryString(context.getContentResolver(), getFileUri(uri), getColumnName());
		if (!TextUtils.isEmpty(url) && !BggContract.INVALID_URL.equals(url)) {
			int i = url.lastIndexOf('/');
			if (i > 0) {
				return url.substring(i + 1);
			}
		}
		return null;
	}
}

File: app/src/main/java/com/boardgamegeek/provider/IndirectFileProvider.kt
package com.boardgamegeek.provider

import android.content.Context
import android.net.Uri
import com.boardgamegeek.extensions.load
import com.boardgamegeek.util.FileUtils

abstract class IndirectFileProvider : BaseFileProvider() {
    protected abstract fun getFileUri(uri: Uri): Uri?

    protected abstract val columnName: String

    override fun generateFileName(context: Context, uri: Uri): String? {
        return getFileUri(uri)?.let {
            val url = context.contentResolver.load(it, arrayOf(columnName))?.use { cursor ->
                if (cursor.count == 1 && cursor.moveToFirst()) {
                    cursor.getString(0)
                } else {
                    null
                }
            }
            FileUtils.getFileNameFromUrl(url)
        }
    }
}
